pkg_downloader/pkg_downloader.re

```diff
@@ -0,0 +1,155 @@
+module Httpkit = Httpkit_lwt_unix_httpaf;
+
+let pkg_root = Fpath.v("./packages");
+let source_file_name = Fpath.v("source.tar.gz");
+
+let skip_unknown_package = pkg => {
+  switch (Model.Package.(pkg.target)) {
+  | `Unknown => false
+  | _ => true
+  };
+};
+
+let skip_already_downloaded_versions = (pkg, version) => {
+  open Model.Package;
+  open Model.Version;
+  let source_dir = Fpath.(pkg_root /\/ v(pkg.name) /\/ v(version.version));
+  let source_path = Fpath.(source_dir /\/ source_file_name);
+  let source_exists = Bos.OS.File.exists(source_path);
+  switch (source_exists) {
+  | Ok(true) =>
+    Logs.info(m => m("Skipping package %s/%s", pkg.name, version.version));
+    false; /* false will skip! */
+  | Ok(false) => true /* true will keep!  */
+  | Error(`Msg(msg)) =>
+    Logs.err(m =>
+      m("Error checking package %s/%s: %s", pkg.name, version.version, msg)
+    );
+    false;
+  };
+};
+
+let pick_newest_major_versions = (name, versions) => {
+  let majors = Hashtbl.create(8);
+  versions
+  |> List.iter(v =>
+       switch (Model.Version.(v.version |> Semver.of_string)) {
```

Comment:
Might make sense to add containers so that we can use `CCOpt.map()` here.

Reply:
I'd be inclined to pick a library like Tablecloth instead, so we can use it on the website too.

It'd be pretty annoying to have to jump from Containers to Belt and back :/ — do you have any thoughts on that?

Reply:
Sounds good. I don't think this is a blocker to getting this merged either.